server/db/driver/pg/upgrades.go

```diff
@@ -0,0 +1,348 @@
+			strconv.FormatFloat(float64(totalVolume)/1e8, 'f', -1, 64), strings.ToUpper(dex.BipIDSymbol(mkt.Base)),
+			strconv.FormatFloat(float64(totalQVolume)/1e8, 'f', -1, 64), strings.ToUpper(dex.BipIDSymbol(mkt.Quote)))
+		return nil
+	}
+
+	for _, mkt := range mkts {
+		err = doMarketMatches(mkt)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// DBVersion retrieves the database version from the meta table.
+func DBVersion(db *sql.DB) (ver uint32, err error) {
+	err = db.QueryRow(internal.SelectDBVersion).Scan(&ver)
+	return
+}
+
+func setDBVersion(db sqlExecutor, ver uint32) error {
+	res, err := db.Exec(internal.SetDBVersion, ver)
+	if err != nil {
+		return err
+	}
+
+	n, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if n != 1 {
+		return fmt.Errorf("set the DB version in %d rows instead of 1", n)
+	}
+	return nil
+}
+
+func upgradeDB(ctx context.Context, db *sql.DB) error {
+	// Get the DB version from the meta table. Nonexistent meta table or
+	// meta.schema_version column implies v0, the upgrade from which adds the
+	// table and schema_version column.
+	var current uint32
+	found, err := tableExists(db, metaTableName)
+	if err != nil {
+		return err
+	}
+	if found {
+		found, err = columnExists(db, "public", metaTableName, "schema_version")
+		if err != nil {
+			return err
+		}
+		if found {
+			current, err = DBVersion(db)
+			if err != nil {
+				return fmt.Errorf("failed to get DB version: %w", err)
+			}
+		} // else v1 upgrade creates meta.schema_version column
+	} // else v1 upgrade creates meta table
+
+	if current == dbVersion {
+		log.Infof("DCRDEX database ready at version %d", dbVersion)
+		return nil // all upgraded
+	}
+
+	if current > dbVersion {
+		return fmt.Errorf("current DB version %d is newer than highest recognized version %d",
+			current, dbVersion)
+	}
+
+	runUpgradeTx := func(targetVer uint32, up func(db *sql.Tx) error) error {
+		// Canceling the context automatically rolls back the transaction.
+		tx, err := db.BeginTx(ctx, nil)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				_ = tx.Rollback() // redundant if ctx canceled (sql.ErrTxDone)
+			}
+		}()
+
+		if err = up(tx); err != nil {
+			return fmt.Errorf("failed to upgrade to db version %d: %w", targetVer, err)
+		}
+
+		if err = setDBVersion(tx, targetVer); err != nil {
+			return fmt.Errorf("failed to set new DB version %d: %w", targetVer, err)
+		}
+
+		err = tx.Commit() // for the defer
+		return err
+	}
+
+	log.Infof("Upgrading DB scheme from %d to %d", current, len(upgrades))
+	for i, up := range upgrades[current:] {
+		targetVer := current + uint32(i) + 1
+		log.Debugf("Upgrading DB scheme to %d...", targetVer)
+		if err = runUpgradeTx(targetVer, up); err != nil {
+			return err
+		}
+	}
+
+	current, err = DBVersion(db)
+	if err != nil {
+		return fmt.Errorf("failed to get DB version: %w", err)
+	}
+	log.Infof("Upgrades complete. DB is at version %d", current)
+	return nil
```

Comment:
Going back to wisdom's comment

> If setting the db version fails, we'll have an upgraded database with an older version number. Can avoid this by using a transaction for each upgrade, pass the transaction to each upgrade function, after the upgrade function returns without error, set the db version before committing. Any error should rollback all changes.

I seems to me like the idea was to use a single transaction for all upgrades and the db version, and not commit the transaction until the db version is recorded successfully. But the wording "using a transaction for each upgrade" seems to conflict with that assumption too, so I guess I don't know. 

But wisdom's initial concern about an upgrade being performed but the version not being recorded has not been addressed, I don't think.

Reply:
I believe it is addressed here: https://github.com/decred/dcrdex/commit/6b4d7c73fb65966372f31e5b7344e47e2cd37aa4#diff-d5886fb14518c5d25332c7350f812301f9c6628f337111aa73258edac37c7472R309-R338

If anything with the actual upgrade or version setting fails, both rollback and you are at a consistent database level.  That it is perhaps at an upgraded version but not the best version isn't a problem.

Reply:
I suppose we could technically wrap the loop of all the upgrades in the same tx like you're thinking, but I'm getting a bit anxious about the limits of postgresql here. I wonder if it's gonna be ok to have say 10 versions all modifying the tables heavily.

Reply:
Ugh, dang I have to re-rig NewDEX for cancellation.  It doesn't work presently to cancel.

Reply:
> I believe it is addressed here: 6b4d7c7#diff-d5886fb14518c5d25332c7350f812301f9c6628f337111aa73258edac37c7472R309-R338

Ah. I see what I did wrong. Carry on.